### ISSUE TITLE
Vertex Shader RCP/RSQ div-by-zero (OOT shading glitches)

### ISSUE THREAD

**Issue by tfarley** [citra-emu/citra#880]
Related to #808

**Issue:** In OOT (other games untested), occasionally an RCP and/or RSQ shader instruction will have a zero or close to zero denominator via `src1[i].ToFloat32()`.

**Effect:** This causes either solid black models or specific white triangles depending on machine/hardware. It also causes the menu selection brackets and gear screen icons to disappear.

**Hacky fix:** Doing an epsilon check on `src1[i].ToFloat32()` that sets `dest[i]` to zero on failure for RCP and RSQ fixes the menu items and models respectively.

Potential sources of the error could include: VFP, float 24<->32 conversion, incorrect data being passed to the shader, or an error elsewhere in the shader code since the calls to RCP/RSQ act on temporary `r` registers in OOT's shader.

**Comment by purpasmart96** [comment]
<code>The RSQ instruction assigns the inverse square root of the
    absolute value of the source scalar into the destination register.

    Additionally, RSQ(0.0) gives [+Inf, +Inf, +Inf, +Inf]; and both
    RSQ(+Inf) and RSQ(-Inf) give [0.0, 0.0, 0.0, 0.0];

        t.x = source0.c;
        if (negate0) {
          t.x = -t.x;
        }
        u.x = 1.0f / sqrt(fabs(t.x));
        if (xmask) destination.x = u.x;
        if (ymask) destination.y = u.x;
        if (zmask) destination.z = u.x;
        if (wmask) destination.w = u.x;

    where

        | u.x - IEEE(1.0f/sqrt(fabs(t.x))) | < 1.0f/(2^22)

    for 1.0f <= t.x <= 4.0f.  The intent of this precision requirement is
    that this amount of relative precision apply over all values of t.x.
</code>

**Comment by mailwl** [comment]
Instead of fixing results of wrong implementition of this opcodes, maybe better fix the implementation itself?
look at example here:
https://msdn.microsoft.com/en-us/library/windows/desktop/bb147346%28v=vs.85%29.aspx

**Comment by taisel** [comment]
I think they're fixing the results as a way to avoid being unable to use the native instruction.

**Comment by mailwl** [comment]
Why native, it citra emulates this?
I mean, we should fix this place in sources:
https://github.com/citra-emu/citra/blob/master/src/video_core/vertex_shader.cpp#L256

**Comment by taisel** [comment]
Well, what's the overhead for doing it bit perfectly? This is of course after working around stuff like x86 80 bit extended precision being on by default non-compliance in a bunch of compiler setups as a speed-up feature. I'm assuming they're trying to emulate the PICA's floating point arith or is this data sent before that?

**Comment by purpasmart96** [comment]
"The RSQ instruction assigns the inverse square root of the
    absolute value of the source scalar into the destination register.

    Additionally, RSQ(0.0) gives [+Inf, +Inf, +Inf, +Inf]; and both
    RSQ(+Inf) and RSQ(-Inf) give [0.0, 0.0, 0.0, 0.0];

        t.x = source0.c;
        if (negate0) {
          t.x = -t.x;
        }
        u.x = 1.0f / sqrt(fabs(t.x));
        if (xmask) destination.x = u.x;
        if (ymask) destination.y = u.x;
        if (zmask) destination.z = u.x;
        if (wmask) destination.w = u.x;

    where

        | u.x - IEEE(1.0f/sqrt(fabs(t.x))) | < 1.0f/(2^22)

    for 1.0f <= t.x <= 4.0f.  The intent of this precision requirement is
    that this amount of relative precision apply over all values of t.x."
source:https://www.opengl.org/registry/specs/NV/vertex_program1_1.txt

**Comment by mailwl** [comment]
here, example of implementation of RSQ, without checking to zero (don't know, which epsilon is enough for float 24).
`
            case OpCode::Id::RSQ:
            {
                float f_x = src1[0].ToFloat32();
                float f_y = src1[1].ToFloat32();
                float f_z = src1[2].ToFloat32();
                float f_w = src1[3].ToFloat32();

                float f_abs = sqrtf(powf(f_x, 2)+powf(f_y, 2)+powf(f_z, 2)+powf(f_w, 2));
                // TODO check f_abs != 0.0
                float24 f_rsq =  float24::FromFloat32(1.0f / f_abs);

                for (int i = 0; i < 4; ++i) {
                    if (!swizzle.DestComponentEnabled(i))
                        continue;
                    dest[i] = f_rsq;
                }

                break;
            }
`

**Comment by taisel** [comment]
The question is, is the PICA following IEEE compliance?

**Comment by purpasmart96** [comment]
@mailwl 
      t.x = source0.c***;
      t.y = source0.*c**;
      t.z = source0.**c*;
      t.w = source0.***c;

    if (xmask) destination.x = u.x;
    if (ymask) destination.y = u.x;
    if (zmask) destination.z = u.x;
    if (wmask) destination.w = u.x;

**Comment by purpasmart96** [comment]
Was Doing some digging around and stumbled upon this
 case D3DSIO_RSQ:
            if(Source.x == 0.0f)
            {
                Result.w = FLT_MAX;
            }
            else
            {
                Result.w = 1.0f / sqrtf(fabsf(Source.x));
            }
            Result.x = Result.y = Result.z = Result.w;

**Comment by neobrain** [comment]
@taisel I'm not sure if IEEE considers 24-bit floats at all. Even if it does, there's only little reason for a GPU to follow strict IEEE rules.

**Comment by mailwl** [comment]
Here, can anyone test it? I have no decrypted roms ATM:
```cpp
            case OpCode::Id::RCP:
            {
                const float FLT_MAX = 3.40282347E+38F;
                float f_x = src1[0].ToFloat32();
                float f_rpc = (f_x==0.0f) ? FLT_MAX : (1.0f / f_x);

                for (int i = 0; i < 4; ++i) {
                    if (!swizzle.DestComponentEnabled(i))
                        continue;
                    dest[i] = float24::FromFloat32(f_rpc);
                }

                break;
            }

            case OpCode::Id::RSQ:
            {
                const float FLT_MAX = 3.40282347E+38F;
                float f_x = src1[0].ToFloat32();
                float f_rsq = (f_x==0.0f) ? FLT_MAX : (1.0f / sqrtf(fabsf(f_x)));

                for (int i = 0; i < 4; ++i) {
                    if (!swizzle.DestComponentEnabled(i))
                        continue;
                    dest[i] = float24::FromFloat32(f_rsq);
                }
                break;
            }
```

**Comment by taisel** [comment]
@neobrain yeah, and that's probably an issue unless the rounding modes and steps are mapped.

**Comment by purpasmart96** [comment]
Is there any way possible that there could we could do some so of HW test for this?

**Comment by Subv** [comment]
@purpasmart96 Of course, just write a shader that uses RSQ/RCP, start feeding it data, and see what comes out the other side

**Comment by mailwl** [comment]
well, i tested oot, still bad. i think, wrong vector component
![zelda1](https://cloud.githubusercontent.com/assets/629437/8311935/6472a10e-19e2-11e5-80bc-7ae7661ce54c.png)
![zelda2](https://cloud.githubusercontent.com/assets/629437/8311937/68acfac6-19e2-11e5-948d-5c5b486ef3a8.png)

**Comment by mailwl** [comment]
sorry, delete previous comment by ocasion.
With fix, i proposed, picture looks better, but not perfect.
Link now mostly green, with many white textures.
I think, hardcoded X coordinate of src is really wrong

**Comment by yuriks** [comment]
@mailwl @taisel as I understand it, the issue isn't really that RCP/RSQ don't work exactly, but that the shader is getting problematic values (for unknown reasons). Checking for 0 is just a workaround.

**Comment by mailwl** [comment]
@yuriks, no! wrong implementation of RCP/RSQ is the problem. Check picture with #808 
![zelda1](https://cloud.githubusercontent.com/assets/629437/8313373/ed124c46-19ea-11e5-9181-644eb0ce8a85.png)

**Comment by yuriks** [comment]
I don't see how that's proof of anything?

**Comment by mailwl** [comment]
with (mostly) right implementation picture looks like on 3ds. But implementation not full correctly

**Comment by purpasmart96** [comment]
This should shed some light on what is happening
<code>
"One of the tell-tale signs of a floating-point special problem is if your application 
periodically has all or part of the screen turn black or white. Since you cannot directly 
visualize a floating-point surface on a GeForce6 GPU, most algorithms dealing with 
floating-point data convert them to fixed-point at some point. In this step, a +Inf 
becomes white (1.0), and –Inf and NaN become black."
</code>
http://http.download.nvidia.com/developer/Papers/2005/FP_Specials/FP_Specials.pdf

**Comment by yuriks** [comment]
Fixed by #948
